android-agent/src/main/java/io/opentelemetry/android/instrumentation/AndroidInstrumentationRegistryImpl.kt

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.android.instrumentation
+
+import java.util.Collections
+import java.util.ServiceLoader
+
+internal class AndroidInstrumentationRegistryImpl : AndroidInstrumentationRegistry {
+    private val instrumentations: MutableMap<Class<out AndroidInstrumentation>, AndroidInstrumentation> by lazy {
+        ServiceLoader.load(AndroidInstrumentation::class.java).associateBy { it.javaClass }
+            .toMutableMap()
```

Comment:
So this means that when an instance of `AndroidInstrumentationRegistryImpl` is created, the map is not yet created, right? As soon as `get()` or `getAll()` or `register()` is called, the map will be created lazily through the `ServiceLoader` mechanism?

Just wanted to think it through. Seems fine, I think, and I like that the user of the class doesn't have to do any work. Pretty elegant.

Reply:
> So this means that when an instance of `AndroidInstrumentationRegistryImpl` is created, the map is not yet created, right? As soon as `get()` or `getAll()` or `register()` is called, the map will be created lazily through the `ServiceLoader` mechanism?

That's right. Tbh I was a bit concerned about doing the service load implicitly as it is now, because I'm not sure if there could be a case where some users might not want to use instrumentations from the classpath and instead register manually the ones they'd like to use, and the way it is now wouldn't support that... But then I didn't want to spend too much time on a use case that I'm not sure will ever be needed, so that's why I left it as is, though I'm curious to know about other opinions on it.

Reply:
I don't think it's worth changing. Let's go with this and adjust as needed.